# 💻 Financial Returns – One-Hour Class Exercise (with Formulas)
In this notebook, you'll practice calculating and comparing different types of returns: **simple return**, **compound return (CAGR)**, and **continuous return**, including the effect of **dividends**.

We'll use Python to implement each calculation step-by-step, and include all formulas in proper math notation using LaTeX.

## 🧮 Problem 1: Interest on $1,000 Invested at 5% for 3 Years

**Given:**
- Initial Investment: $1,000
- Annual Interest Rate: 5% (0.05)
- Time: 3 years

Solve for:
- Simple Return
- Future Value with Periodic (Annual) Compounding
- Future Value with Continuous Compounding

### 🔹 Simple Interest Formula:
$$
FV = PV \times (1 + rt)
$$
- $PV$ = Principal (Initial Investment)
- $r$ = Annual interest rate
- $t$ = Number of years

In [ ]:
# Inputs
PV = 1000
r = 0.05
t = 3

# Simple interest
FV_simple = PV * (1 + r * t)
print(f"Future Value with Simple Interest: ${FV_simple:.2f}")
print("Interpretation: You earn $150 interest over 3 years using simple interest.")

Future Value with Simple Interest: $1150.00
Interpretation: You earn $150 interest over 3 years using simple interest.


### 🔹 Periodic Compounding (Annual)
$$
FV = PV \times \left(1 + \frac{r}{n}\right)^{nt}
$$
- $n$ = number of compounding periods per year (annually = 1)

In [ ]:
n = 1
FV_periodic = PV * (1 + r/n)**(n * t)
print(f"Future Value with Annual Compounding: ${FV_periodic:.2f}")
print("Interpretation: Compounding annually gives you slightly more interest ($157.63) than simple interest.")

Future Value with Annual Compounding: $1157.63
Interpretation: Compounding annually gives you slightly more interest ($157.63) than simple interest.


### 🔹 Continuous Compounding
$$
FV = PV \times e^{rt}
$$
Where $e \approx 2.718$ is Euler’s number.

In [ ]:
import numpy as np
FV_continuous = PV * np.exp(r * t)
print(f"Future Value with Continuous Compounding: ${FV_continuous:.2f}")
print("Interpretation: Continuous compounding yields the highest return ($161.83) due to constant reinvestment.")

Future Value with Continuous Compounding: $1161.83
Interpretation: Continuous compounding yields the highest return ($161.83) due to constant reinvestment.


### 📊 Summary of Results
| Method                  | Future Value | Interest Earned |
|-------------------------|--------------|------------------|
| Simple Interest         | $1,150.00    | $150.00          |
| Periodic (Annual)       | $1,157.63    | $157.63          |
| Continuous Compounding  | $1,161.83    | $161.83          |

**Conclusion:** The more frequently interest is compounded, the greater the return. Continuous compounding is a theoretical maximum.

## 🧮 Problem 2:

**Given:**
- Initial Investment: $2,000

- Final Investment: $2,640

- Dividens: $50

- Years invested: 4

Solve for:
- Simple Return
- Future Value with Periodic (Annual) Compounding
- Future Value with Continuous Compounding

## 1️⃣ Simple Return with Dividends
### 📌 Formula:
$$
\text{Simple Return} = \frac{P_{\text{final}} + D - P_{\text{initial}}}{P_{\text{initial}}}
$$
**Where:**
- $P_{\text{initial}}$ = Initial investment
- $P_{\text{final}}$ = Final value
- $D$ = Total dividends received

In [ ]:
# Inputs
initial_value = 2000
final_value = 2640
dividends = 50

# Simple return with dividends
simple_return = (final_value + dividends - initial_value) / initial_value
print(f"Simple Return with Dividends: {simple_return:.4f} or {simple_return*100:.2f}%")

Simple Return with Dividends: 0.3450 or 34.50%


## 2️⃣ Compound Annual Growth Rate (CAGR)
### 📌 Formula:
$$
\text{CAGR} = \left(\frac{P_{\text{final}} + D}{P_{\text{initial}}}\right)^{\frac{1}{t}} - 1
$$
**Where:**
- $t$ = Number of years

In [ ]:
# Years invested
t = 4

# CAGR calculation
cagr = ((final_value + dividends) / initial_value) ** (1 / t) - 1
print(f"CAGR: {cagr:.4f} or {cagr*100:.2f}% annually")

CAGR: 0.0769 or 7.69% annually


## 3️⃣ Continuous (Log) Return
### 📌 Formula:
$$
\text{Log Return} = \ln\left(\frac{P_{\text{final}} + D}{P_{\text{initial}}}\right)
$$
### To annualize:
$$
\text{Annual Log Return} = \frac{\text{Log Return}}{t}
$$

In [ ]:
# Log return calculation
log_return = np.log((final_value + dividends) / initial_value)
annual_log_return = log_return / t

print(f"Total Log Return: {log_return:.4f} or {log_return*100:.2f}%")
print(f"Annual Log Return: {annual_log_return:.4f} or {annual_log_return*100:.2f}%")

Total Log Return: 0.2964 or 29.64%
Annual Log Return: 0.0741 or 7.41%
